File: android-tv/app/src/main/java/com/donghuaflix/ui/detail/DetailViewModel.kt
package com.donghuaflix.ui.detail

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.donghuaflix.data.repository.ShowRepository
import com.donghuaflix.data.repository.WatchRepository
import com.donghuaflix.domain.model.Episode
import com.donghuaflix.domain.model.Show
import com.donghuaflix.domain.model.WatchProgress
import com.donghuaflix.domain.model.WebsiteInfo
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class EpisodeWatchInfo(
    val episodeNumber: Int,
    val progressFraction: Float, // 0.0 to 1.0
    val completed: Boolean,
)

data class DetailUiState(
    val show: Show? = null,
    val episodes: List<Episode> = emptyList(),
    val selectedWebsite: WebsiteInfo? = null,
    val lastWatched: WatchProgress? = null,
    val isInWatchlist: Boolean = false,
    val isLoading: Boolean = true,
    val watchedEpisodes: Map<Int, EpisodeWatchInfo> = emptyMap(),
)

@HiltViewModel
class DetailViewModel @Inject constructor(
    savedStateHandle: SavedStateHandle,
    private val showRepository: ShowRepository,
    private val watchRepository: WatchRepository,
) : ViewModel() {

    private val showId: Int = savedStateHandle["showId"] ?: 0

    private val _uiState = MutableStateFlow(DetailUiState())
    val uiState: StateFlow<DetailUiState> = _uiState.asStateFlow()

    init {
        loadShow()
    }

    private fun loadShow() {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true) }

            val show = showRepository.getShow(showId)
            val inWatchlist = watchRepository.isInWatchlist(showId)
            val lastWatched = watchRepository.getLastWatchedForShow(showId)

            // Fetch all watched episodes for this show
            val watchHistory = watchRepository.getWatchedEpisodesForShow(showId)
            val watchedEpisodes = watchHistory.mapValues { (_, wp) ->
                val fraction = if (wp.durationSeconds != null && wp.durationSeconds > 0) {
                    wp.progressSeconds.toFloat() / wp.durationSeconds
                } else 0f
                EpisodeWatchInfo(
                    episodeNumber = wp.episodeNumber,
                    progressFraction = fraction,
                    completed = wp.completed,
                )
            }

            if (show != null) {
                val defaultWebsite = show.websites.firstOrNull()
                _uiState.update {
                    it.copy(
                        show = show,
                        selectedWebsite = defaultWebsite,
                        lastWatched = lastWatched,
                        isInWatchlist = inWatchlist,
                        isLoading = false,
                        watchedEpisodes = watchedEpisodes,
                    )
                }
                // Load episodes for default website
                if (defaultWebsite != null) {
                    loadEpisodes(defaultWebsite.name)
                }
            } else {
                _uiState.update { it.copy(isLoading = false) }
            }
        }
    }

    fun selectWebsite(website: WebsiteInfo) {
        _uiState.update { it.copy(selectedWebsite = website) }
        loadEpisodes(website.name)
    }

    private fun loadEpisodes(websiteName: String) {
        viewModelScope.launch {
            val episodes = showRepository.getEpisodes(showId, websiteName)
            _uiState.update { it.copy(episodes = episodes) }
        }
    }

    fun refreshWatchHistory() {
        viewModelScope.launch {
            val lastWatched = watchRepository.getLastWatchedForShow(showId)
            val watchHistory = watchRepository.getWatchedEpisodesForShow(showId)
            val watchedEpisodes = watchHistory.mapValues { (_, wp) ->
                val fraction = if (wp.durationSeconds != null && wp.durationSeconds > 0) {
                    wp.progressSeconds.toFloat() / wp.durationSeconds
                } else 0f
                EpisodeWatchInfo(
                    episodeNumber = wp.episodeNumber,
                    progressFraction = fraction,
                    completed = wp.completed,
                )
            }
            _uiState.update {
                it.copy(
                    lastWatched = lastWatched,
                    watchedEpisodes = watchedEpisodes,
                )
            }
        }
    }

    fun markEpisodeWatched(episodeNumber: Int) {
        viewModelScope.launch {
            watchRepository.markAsWatched(showId, episodeNumber)
            refreshWatchHistory()
        }
    }

    fun markEpisodeUnwatched(episodeNumber: Int) {
        viewModelScope.launch {
            watchRepository.markAsUnwatched(showId, episodeNumber)
            refreshWatchHistory()
        }
    }

    fun markAllWatchedUpTo(episodeNumber: Int) {
        viewModelScope.launch {
            val episodes = _uiState.value.episodes.filter { it.episodeNumber <= episodeNumber }
            for (ep in episodes) {
                watchRepository.markAsWatched(showId, ep.episodeNumber)
            }
            refreshWatchHistory()
        }
    }

    fun toggleWatchlist() {
        viewModelScope.launch {
            watchRepository.toggleWatchlist(showId)
            _uiState.update { it.copy(isInWatchlist = !it.isInWatchlist) }
        }
    }
}

File: apple/DonghuaFlix/Features/Detail/DetailViewModel.swift
import Foundation
import Observation

@Observable
final class DetailViewModel {

    // MARK: - State

    var show: Show?
    var episodes: [Episode] = []
    var selectedWebsite: WebsiteInfo?
    var lastWatched: WatchProgress?
    var isInWatchlist = false
    var watchedEpisodes: [Int: WatchProgress] = [:]
    var preloadingEpisodes: Set<Int> = []
    var preloadedEpisodes: Set<Int> = []
    var isLoading = false
    var latestFirst = true
    var error: String?

    // Episode paging (for 50+ episodes)
    var currentPage = 0
    private let episodesPerPage = 50

    var totalPages: Int {
        max(1, Int(ceil(Double(sortedEpisodes.count) / Double(episodesPerPage))))
    }

    var sortedEpisodes: [Episode] {
        latestFirst ? episodes.sorted { $0.episodeNumber > $1.episodeNumber } : episodes.sorted { $0.episodeNumber < $1.episodeNumber }
    }

    var pagedEpisodes: [Episode] {
        let sorted = sortedEpisodes
        let start = currentPage * episodesPerPage
        let end = min(start + episodesPerPage, sorted.count)
        guard start < sorted.count else { return [] }
        return Array(sorted[start..<end])
    }

    var nextEpisodeNumber: Int? {
        if let lastWatched = lastWatched {
            if lastWatched.completed {
                return lastWatched.episodeNumber + 1
            }
            return lastWatched.episodeNumber
        }
        return episodes.min(by: { $0.episodeNumber < $1.episodeNumber })?.episodeNumber
    }

    private let showRepository: ShowRepository
    private let watchRepository: WatchRepository

    init(showRepository: ShowRepository, watchRepository: WatchRepository) {
        self.showRepository = showRepository
        self.watchRepository = watchRepository
    }

    // MARK: - Load

    @MainActor
    func loadShow(id: Int) async {
        guard !isLoading else { return }
        isLoading = true
        error = nil

        do {
            // Fetch show details
            show = try await showRepository.getShow(id: id)

            // Select first website by default
            if let firstWebsite = show?.websites.first, selectedWebsite == nil {
                selectedWebsite = firstWebsite
            }

            // Fetch episodes for selected website
            await loadEpisodes(showId: id)

            // Fetch watch history
            watchedEpisodes = try await watchRepository.getWatchedEpisodesForShow(showId: id)

            // Find last watched episode
            lastWatched = watchedEpisodes.values
                .sorted(by: { $0.episodeNumber > $1.episodeNumber })
                .first

            // Check watchlist
            isInWatchlist = watchRepository.isInWatchlist(showId: id)
        } catch {
            self.error = error.localizedDescription
        }

        isLoading = false
    }

    @MainActor
    func selectWebsite(_ website: WebsiteInfo) async {
        guard website.id != selectedWebsite?.id else { return }
        selectedWebsite = website
        if let showId = show?.id {
            await loadEpisodes(showId: showId)
        }
    }

    @MainActor
    private func loadEpisodes(showId: Int) async {
        do {
            episodes = try await showRepository.getEpisodes(
                showId: showId,
                website: selectedWebsite?.name
            )
        } catch {
            self.error = error.localizedDescription
        }
    }

    // MARK: - Watch Actions

    @MainActor
    func markEpisodeWatched(_ episode: Episode) async {
        guard let showId = show?.id else { return }
        do {
            try await watchRepository.markAsWatched(showId: showId, episodeNumber: episode.episodeNumber, episodeId: episode.id)
            watchedEpisodes[episode.episodeNumber] = WatchProgress(
                showId: showId,
                episodeNumber: episode.episodeNumber,
                progressSeconds: 0,
                durationSeconds: 0,
                completed: true
            )
        } catch {
            // Silently fail
        }
    }

    @MainActor
    func markEpisodeUnwatched(_ episode: Episode) async {
        guard let showId = show?.id else { return }
        do {
            try await watchRepository.markAsUnwatched(showId: showId, episodeNumber: episode.episodeNumber, episodeId: episode.id)
            watchedEpisodes.removeValue(forKey: episode.episodeNumber)
        } catch {
            // Silently fail
        }
    }

    @MainActor
    func markAllWatchedUpTo(_ episodeNumber: Int) async {
        guard let showId = show?.id else { return }
        for episode in episodes where episode.episodeNumber <= episodeNumber {
            if watchedEpisodes[episode.episodeNumber]?.completed != true {
                do {
                    try await watchRepository.markAsWatched(showId: showId, episodeNumber: episode.episodeNumber, episodeId: episode.id)
                    watchedEpisodes[episode.episodeNumber] = WatchProgress(
                        showId: showId,
                        episodeNumber: episode.episodeNumber,
                        progressSeconds: 0,
                        durationSeconds: 0,
                        completed: true
                    )
                } catch {
                    continue
                }
            }
        }
    }

    @MainActor
    func toggleWatchlist() async {
        guard let showId = show?.id else { return }
        do {
            try await watchRepository.toggleWatchlist(showId: showId)
            isInWatchlist = watchRepository.isInWatchlist(showId: showId)
        } catch {
            // Silently fail
        }
    }

    // MARK: - Preload Sources

    @MainActor
    func preloadSources(for episode: Episode) async {
        guard !preloadingEpisodes.contains(episode.episodeNumber),
              !preloadedEpisodes.contains(episode.episodeNumber) else { return }

        preloadingEpisodes.insert(episode.episodeNumber)

        do {
            let sources = try await showRepository.getEpisodeSources(episodeId: episode.id)
            if !sources.isEmpty {
                preloadedEpisodes.insert(episode.episodeNumber)
            }
        } catch {
            // Silently fail
        }

        preloadingEpisodes.remove(episode.episodeNumber)
    }
}
